### DISCUSSION TITLE
MixedCrossGradCurlIntegrator in 2D?

### DISCUSSION THREAD

**Discussion by M8kmyday** [mfem/mfem#2528] [Q&A] [answered]
MixedCrossGradCurlIntegrator is available in 3D, but I need it in 2D, so I am digging in to see if I can create a 2D version.  If this already been done or if someone has hints, I would greatly appreciate hearing about it.

**Comment by tzanio** [comment]
ping: @mlstowell

**Comment by M8kmyday** [comment, marked as answer]
Modifying  MixedCrossGradCurlIntegrator looks pretty daunting.  I went back to the PDE and re-formulated it to be able to use the existing 2D operators MixedGradGradIntegrator and MixedVectorWeakDivergenceIntegrator.  In terms of MixedCrossGradCurlIntegrator, lambda is the unit vector in the z-direction, and delta is defined only over the x- and y-dimensions.  Working through the math then allows the simplification.

**Comment by mlstowell** [comment]
Does your lambda vector have a component in the z-direction? If its strictly in the xy-plane or strictly in the z-direction the operator would be simplified in different ways. We should try to support both of these if they don't already exist.

**Comment by M8kmyday** [comment]
Lambda is in the z-direction, while the 2D problem is in the x-y plane.  The screen shot below shows the formula of interest as the second equation in (5).  Eq. (4) shows how 3D terms are broken into 2D and z terms.
![Screenshot from 2021-09-14 14-31-24](https://user-images.githubusercontent.com/90462987/133322497-8040d9b1-39fc-495d-9aa1-d89f74cf968f.png)
In this electromagnetic problem, the goal is to calculate the eigenvalue problem to find the propagation constants and the 2D field structures of the modes on a transmission line or waveguide.  This is a 2D problem with the assumed characteristic of propagation on the guiding structure.

**Reply by mlstowell** [reply]
Hi, @M8kmyday, 

I see. So you'll be using `E_t` in H(Curl) and `E_z` in H1. The second equation in (5) involves three bilinear forms (and a slight abuse of notation since the left side is a vector and the right side is a scalar). The one on the right hand side is clearly an H1 mass matrix with `epsilon_r` as a coefficient (that's the easy one). The first operator on the left hand side involves `E_z` and this simplifies to `-Div(mu^{-1} Grad(E_z)) z_hat`. This is just a `DiffusionIntegrator` in 2D if you ignore the vector `z_hat` (as we should). The final piece involves `E_t` but this reduces to a weak divergence so you can use `MixedVectorWeakDivergenceIntegrator`.

Best wishes,
Mark

**Comment by M8kmyday** [comment]
Thanks.

**Comment by M8kmyday** [comment]
I am working with
- ex25p for complex numbers
- ex5p for solving two coupled PDEs with BlockVector
- ex13p for the eigenvalue setup
I'm finding that I am having to dig deep into basic definitions and functions to create a setup the solves two coupled PDEs with complex variables.  I'm afraid that that will take me an extremely long time to work out.  Is there an additional example or pseudo code that demonstrates the basic setup for this?  Something with the natural boundary conditions would be fine.

**Comment by mlstowell** [comment]
Hello, @M8kmyday,

The complex classes are a relatively new part of MFEM and we don't have many examples showing their usage. However, they are little more than convenience classes which allow you to treat the real and imaginary parts of fields and operators as single entities. I'm certain that MFEM can produce all of the real-valued operators required to setup your complex-valued problem. I believe we also have the capability to combine these operators into a single large CSR matrix for passing along to an eigenvalue solver.

Speaking of eigenvalue solvers; I don't think `ex13p` is going to offer much help. The eigenvalue solver used in that example is rather specific to the standard Maxwell eigenvalue problem. In the past I have used ARPACK for similar problems but there are also more modern alternatives such as SLEPc or FEAST. SLEPc should be available through MFEM's PETSc interface. Other options would require linking with third party libraries and figuring out the matrix format they require.

I'd be happy to advise you on building the appropriate operators with or without essential boundary conditions and producing the correct matrix format for your chosen eigenvalue solver. Unfortunately, MFEM just doesn't yet have an end-to-end solution for this type of problem.

Best wishes,
Mark

**Comment by M8kmyday** [comment]
Good info.  Thanks.

I'm certainly interested in advice on using MFEM.  Can we take this to email?  I can provide some background on the goals of my project, which in short, is an open-source EM simulator to provide the functionality of HFSS.  My email is briany at austin.rr.com.

Brian
